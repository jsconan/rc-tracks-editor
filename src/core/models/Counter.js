/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { eventEmitterMixin } from '../mixins';

/**
 * Represents a set of counters.
 */
export class Counter extends Map {
    /**
     * Creates an observable set of counters.
     * Each change in the counters will be notified to the subscribers.
     * @param {*} source - An iterable object that can be used to initialize the counters.
     */
    constructor(source = null) {
        super();

        eventEmitterMixin(this);

        if (source) {
            this.load(source);
            source = void 0;
        }
    }

    /**
     * Applies a callback to each counter from the list.
     * @param {function} walker - A callback that will be applied to each counter of the list.
     * @returns {Counter} - Chains the instance.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        super.forEach((value, key) => walker.call(this, value, key, this));

        return this;
    }

    /**
     * Maps the counters to an object.
     * @param {function} mapper - A mapper callback that will be applied to each counter.
     * @returns {object}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = mapper.call(this, value, key, this);
        }

        return counters;
    }

    /**
     * Gets the value for a particular counter.
     * @param {string} key - The key of the counter to read.
     * @returns {number} - The current value of the counter. It returns 0 if the counter does not exist.
     */
    get(key) {
        return super.get(key) || 0;
    }

    /**
     * Sets the value for a particular counter.
     * @param {string} key - The key of the counter to set.
     * @param {number} value  - The new value of the counter. It must be a whole number.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    set(key, value) {
        const previous = super.get(key) || 0;
        super.set(key, Math.floor(value) || 0);

        this.emit('set', key, value, previous);

        return this;
    }

    /**
     * Removes a counter.
     * @param {string} key - The key of the counter to delete.
     * @returns {boolean} - Returns `true` if a counter was deleted.
     * @fires delete
     */
    delete(key) {
        const value = super.get(key) || 0;
        const deleted = super.delete(key);

        if (deleted) {
            this.emit('delete', key, value);
        }

        return deleted;
    }

    /**
     * Increments a counter.
     * @param {string} key - The key of the counter to increment.
     * @param {number} amount - The amount to add to the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    increment(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) + value);
    }

    /**
     * Decrements a counter.
     * @param {string} key - The key of the counter to decrement.
     * @param {number} amount - The amount to subtract from the counter.
     * @returns {Counter} - Chains the instance.
     * @fires set
     */
    decrement(key, amount = 1) {
        const value = Math.floor(amount) || 1;
        return this.set(key, this.get(key) - value);
    }

    /**
     * Removes all counters.
     * @returns {Counter} - Chains the instance.
     * @fires clear
     */
    clear() {
        super.clear();

        this.emit('clear');

        return this;
    }

    /**
     * Loads counter from another source. The counters are cleared before.
     * @param {*} iterator - An iterable object that can be used to set the counters.
     * @returns {Counter} - Chains the instance.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        super.clear();
        for (const [key, value] of iterator) {
            super.set(key, Math.floor(value) || 0);
        }

        this.emit('load');

        return this;
    }

    /**
     * Exports the counters to an object.
     * @returns {object} - An object containing the counters.
     */
    toObject() {
        const counters = {};

        for (const [key, value] of this.entries()) {
            counters[key] = value;
        }

        return counters;
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * Notifies a value has been set to a counter.
 * @event set
 * @param {string} key - The key of the counter that was set.
 * @param {number} newValue - The new value.
 * @param {number} oldValue - The previous value.
 */

/**
 * Notifies a counter has been removed.
 * @event delete
 * @param {string} key - The key of the counter that was removed.
 * @param {number} value - The last value of the removed counter.
 */

/**
 * Notifies all counters has been removed.
 * @event clear
 */

/**
 * Notifies the counters has been loaded.
 * @event load
 */
