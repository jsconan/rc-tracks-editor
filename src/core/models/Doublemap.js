/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';

/**
 * Defines a doublemap structure.
 *
 * A doublemap is a `Map` where each value is indexed by two keys, a primary and a secondary. Internally, it
 * relies on a combination of `Map` objects. The primary keys are managed by a first `Map`, then each key
 * is associated with a another `Map` object that contains the secondary keys and the related values. In the
 * end, it could be seen as a two-dimensional associative array.
 */
export class Doublemap extends Map {
    /**
     * Creates a doublemap.
     * @param {*} source - A list of primaryKey/secondaryKey/value sets to inject in the map.
     */
    constructor(source) {
        super();

        if (source) {
            for (const [primaryKey, secondaryKey, value] of source) {
                this.set(primaryKey, secondaryKey, value);
            }
        }
    }

    /**
     * The amount of secondary keys registered in the map.
     * @type {number}
     */
    get size2() {
        let size = 0;
        for (const bucket of super.values()) {
            size += bucket.size;
        }
        return size;
    }

    /**
     * Tells whether a key exists or not.
     * It accepts either the primary key only, or the two keys, or the two keys and the value.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @param {*} [value] - An optional value to check together with the pair of keys.
     * @returns {boolean} - Returns `true` if the keys exist.
     */
    has(primaryKey, secondaryKey, value) {
        const undefinedValue = 'undefined' === typeof value;
        if ('undefined' === typeof secondaryKey && undefinedValue) {
            return super.has(primaryKey);
        }

        const bucket = super.get(primaryKey);
        if (!bucket) {
            return false;
        }

        if (undefinedValue) {
            return bucket.has(secondaryKey);
        }

        return bucket.get(secondaryKey) === value;
    }

    /**
     * Gets the value registered under the given pair of keys.
     * If the value does not exist, `undefined` will be returned.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @returns {*} - Returns either the secondary `Map`, the registered value, or `undefined`.
     */
    get(primaryKey, secondaryKey) {
        const bucket = super.get(primaryKey);

        if (!bucket || 'undefined' === typeof secondaryKey) {
            return bucket;
        }

        return bucket.get(secondaryKey);
    }

    /**
     * Sets a value under a pair of keys.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} secondaryKey - The secondary key to check.
     * @param {*} value - The value to register under the keys.
     * @returns {Doublemap} - Chains the instance.
     */
    set(primaryKey, secondaryKey, value) {
        if ('undefined' !== typeof secondaryKey) {
            let bucket = super.get(primaryKey);
            if (!bucket) {
                bucket = new Map();
                super.set(primaryKey, bucket);
            }
            bucket.set(secondaryKey, value);
        }
        return this;
    }

    /**
     * Deletes a key. If only the primary key is supplied, all its secondary keys will be deleted.
     * If all keys are supplied, only the exact pair will be deleted.
     * If the value is also supplied, it will only be removed if it matches the registered value.
     * @param {*} primaryKey - The primary key to check.
     * @param {*} [secondaryKey] - The secondary key to check.
     * @param {*} [value] - An optional value to remove from the key.
     * @returns {boolean} - Returns `true` if something was deleted.
     */
    delete(primaryKey, secondaryKey, value) {
        const undefinedValue = 'undefined' === typeof value;
        if ('undefined' === typeof secondaryKey && undefinedValue) {
            return super.delete(primaryKey);
        }

        const bucket = super.get(primaryKey);
        if (!bucket) {
            return false;
        }

        if (undefinedValue || bucket.get(secondaryKey) === value) {
            const deleted = bucket.delete(secondaryKey);

            if (!bucket.size) {
                super.delete(primaryKey);
            }

            return deleted;
        }

        return false;
    }

    /**
     * Calls a function on every primaryKey/secondaryKey/value sets of the map.
     * @param {eachKeysValue} each - A callback called for each keys/value pair.
     * @param {*} [thisArg] - An optional call context for the callback. If omitted it will be defaulted to `undefined`.
     * @throws {TypeError} - If the callback is not supplied or is not a function.
     */
    forEach(each, thisArg) {
        validateCallback(each);

        for (const [primaryKey, secondaryKey, value] of this.entries()) {
            each.call(thisArg, value, secondaryKey, primaryKey, this);
        }
    }

    /**
     * Produces an iterator to loop over all the keys.
     * If the primary key is supplied, the iterator will list its secondary keys if any.
     * @param {*} [key] - The primary key for which list the secondary keys.
     * @yields {*} - The next value in the map.
     * @generator
     */
    *keys(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.keys();
            }
            return;
        }

        for (const [primaryKey, bucket] of super.entries()) {
            for (const secondaryKey of bucket.keys()) {
                yield [primaryKey, secondaryKey];
            }
        }
    }

    /**
     * Produces an iterator to loop over all the values.
     * If the primary key is supplied, the iterator will list its values.
     * Otherwise, it will loops through all secondary map as well.
     * @param {*} [key] - The primary key for which list the values.
     * @yields {*} -  The next value in the map.
     * @generator
     */
    *values(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.values();
            }
            return;
        }

        for (const bucket of super.values()) {
            yield* bucket.values();
        }
    }

    /**
     * Produces an iterator to loop over all the primaryKey/secondaryKey/value sets.
     * If the primary key is supplied, the iterator will list its key/value pairs.
     * @param {*} [key] - The primary key for which list the values.
     * @yields {*} - The next primaryKey/secondaryKey/value set in the map.
     * @generator
     */
    *entries(key) {
        if ('undefined' !== typeof key) {
            const bucket = super.get(key);
            if (bucket) {
                yield* bucket.entries();
            }
            return;
        }

        for (const [primaryKey, bucket] of super.entries()) {
            for (const [secondaryKey, value] of bucket) {
                yield [primaryKey, secondaryKey, value];
            }
        }
    }

    /**
     * Produces an iterator to loop over all the primaryKey/secondaryKey/value sets.
     * @yields {*} - The next primaryKey/secondaryKey/value set in the map.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.entries();
    }
}

/**
 * Represents a callback function that is invoked for each primaryKey/secondaryKey/value set of a `Doublemap`
 * object. Mainly used when applying a traversal algorithm, like `forEach()`.
 * @param {*} value - The current value being traversed.
 * @param {*} secondaryKey - The current secondaryKey being traversed.
 * @param {*} primaryKey - The current primaryKey being traversed.
 * @param {Doublemap} map - The `Doublemap` instance being traversed.
 * @callback eachKeysValue
 */
