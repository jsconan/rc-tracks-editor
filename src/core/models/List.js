/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { eventEmitterMixin } from '../mixins';

/**
 * Represents an observable list.
 * Each change in the list will be notified to the subscribers.
 * @mixes EventEmitter
 */
export class List {
    /**
     * Creates an observable list.
     * Each change in the list will be notified to the subscribers.
     * @param {*} source - An iterable object that can be used to initialize the list.
     */
    constructor(source = []) {
        eventEmitterMixin(this);

        this.list = [...source];
        source = void 0;
    }

    /**
     * The number of values in the list.
     * @type {number}
     */
    get length() {
        return this.list.length;
    }

    /**
     * Iterates over the values from the list.
     * @yields {*} - The next value in the list.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.list.values();
    }

    /**
     * Returns an iterator for the values from the list.
     * @yields {*} - The next value in the list.
     * @generator
     */
    *values() {
        yield* this.list.values();
    }

    /**
     * Applies a callback to each value from the list.
     * @param {listCallback} walker - A callback that will be applied to each value of the list.
     * @returns {List} - Chains the list.
     * @throws {TypeError} - If the given callback is not a function.
     */
    forEach(walker) {
        validateCallback(walker);

        this.list.forEach((value, index) => walker.call(this, value, index, this));

        return this;
    }

    /**
     * Maps the values of the list to an array.
     * @param {listCallback} mapper - A mapper callback that will be applied to each value of the list.
     * @returns {Array}
     * @throws {TypeError} - If the given callback is not a function.
     */
    map(mapper) {
        validateCallback(mapper);

        return this.list.map((value, index) => mapper.call(this, value, index, this));
    }

    /**
     * Finds the index of a value from the list. If the values does not exist in the list, `-1` is returned.
     * @param {listCallback|*} filter - Either a filter callback that will be applied to each value of the list
     * and that must return true when the received value matches, or the searched value.
     * @returns {number} - The index of the value in the list, or -1 if not found.
     */
    find(filter) {
        if ('function' === typeof filter) {
            return this.list.findIndex((value, index) => filter.call(this, value, index, this));
        }

        return this.list.indexOf(filter);
    }

    /**
     * Gets the value at a particular index.
     * @param {number} index - The index of the value to get
     * @returns {*} - The value at the given index. It may be undefined.
     */
    get(index) {
        return this.list[index];
    }

    /**
     * Sets a value at a particular index.
     * @param {number} index - The index where to set the value.
     * @param {*} value - The value to set at the index.
     * @returns {List} - Chains the list.
     * @fires set
     * @throws {ReferenceError} - If the given index is out of bounds.
     */
    set(index, value) {
        if (index < 0 || index >= this.length) {
            throw new ReferenceError('The list index is out of bounds!');
        }

        const previous = this.list[index];
        this.list[index] = value;

        /**
         * Notifies a value has been set to the list.
         * @event set
         * @param {number} index - The index where the value was set.
         * @param {*} newValue - The new value.
         * @param {*} oldValue - The previous value.
         */
        this.emit('set', index, value, previous);

        return this;
    }

    /**
     * Inserts a value at a particular index.
     * @param {number} index - The index where to insert the value.
     * @param {...*} value - The value to insert at the index.
     * @returns {List} - Chains the list.
     * @fires insert
     */
    insert(index, ...value) {
        this.list.splice(index, 0, ...value);

        /**
         * Notifies values have been inserted into the list.
         * @event insert
         * @param {number} index - The index where the values have been inserted.
         * @param {...*} value - The inserted values.
         */
        this.emit('insert', index, ...value);

        return this;
    }

    /**
     * Adds a value at the end of the list.
     * @param {...*} value - The value to add.
     * @returns {List} - Chains the list.
     * @fires add
     */
    add(...value) {
        this.list.push(...value);

        /**
         * Notifies values have been added to the list.
         * @event add
         * @param {...*} value - The added value.
         */
        this.emit('add', ...value);

        return this;
    }

    /**
     * Removes a value from the given index.
     * @param {number} index - The index from where remove the value.
     * @param {number} count - The number of values to remove from the index.
     * @returns {number} - The number of deleted values.
     * @fires delete
     */
    delete(index, count = 1) {
        const removed = this.list.splice(index, count);
        const deleted = removed.length;

        if (deleted) {
            /**
             * Notifies values have been removed from the list.
             * @event delete
             * @param {number} index - The index from where the values were removed.
             * @param {...*} value - The removed value.
             */
            this.emit('delete', index, ...removed);
        }

        return deleted;
    }

    /**
     * Clears the list.
     * @returns {List} - Chains the list.
     * @fires clear
     */
    clear() {
        this.list = [];

        /**
         * Notifies the list was cleared.
         * @event clear
         */
        this.emit('clear');

        return this;
    }

    /**
     * Loads values from another source. The list is cleared before.
     * @param {*} iterator - An iterable object that can be used to fill the list.
     * @returns {List} - Chains the list.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        this.list = [...iterator];

        /**
         * Notifies the list was loaded.
         * @event load
         */
        this.emit('load');

        return this;
    }

    /**
     * Copy the list to an array.
     * @returns {Array} - An array containing the values from the list, in the same order.
     */
    toArray() {
        return [...this.list];
    }

    /**
     * Validates that the given object is an instance of the class.
     * Otherwise, an error is thrown.
     * @param {object} object - The instance to validate.
     * @throws {TypeError} - If the given object is not a valid instance.
     */
    static validateInstance(object) {
        if (!(object instanceof this)) {
            throw new TypeError(`The object must be an instance of ${this.name}!`);
        }
    }
}

/**
 * Callback called from the iteration algorithms.
 * @param {*} value - The current value being traversed.
 * @param {number} index - The index of the current value being traversed.
 * @returns {*} - Returns a value expected by the context.
 * @callback listCallback
 */
