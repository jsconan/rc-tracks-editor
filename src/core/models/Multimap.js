/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';

/**
 * Defines a multimap structure.
 *
 * A multimap is a `Map` where each key can have multiple values. Internally, each key receives a bucket that
 * contains the values. This bucket is represented by a `Set`, which contains the list of values registered
 * under the key.
 *
 * When a value is set to a key, a bucket is created if needed, and the value is added to it.
 * Each further set to the same key will add values to the bucket.
 *
 * When a value is deleted from a key, the value is removed from the bucket bound to the key.
 * If the bucket is empty, the key is also removed.
 *
 * When a key is deleted, the bucket is deleted, removing all registered values as well.
 *
 * The iterators work as follow, considering a multimap such as `{ one: [1, 2], two:[3, 4] }`:
 * - `keys()` allows to loop over all the unique keys registered.
 *
 *   Say with the above example: `['one', 'two']`
 *
 * - `values()z allows to loop over all the values registered, one by one.
 *
 *   Say with the above example: `[1, 2, 3, 4]`
 *
 * - `entries()` allows to loop over all the unique key/value pairs.
 *
 *   Say with the above example: `[['one', 1], ['one', 2], ['two', 3], ['two', 4]]`
 *
 * - Exporting a multimap to an array, like with `[...multimap]`, will produce an array of key/value pairs.
 *
 *   Say with the above example: `[['one', 1], ['one', 2], ['two', 3], ['two', 4]]`
 *
 * A multimap can be initialized from either an array or another instance. When an array is used, it must be an array
 * of key/value pairs. To assign multiple values to a key, the key can be duplicated in several pairs, for example:
 * `[['one', 1], ['one', 2], ['two', 3], ['two', 4]]`.
 *
 * The `get()` method will give direct access to the bucket used to store values.
 */
export class Multimap extends Map {
    /**
     * Creates a multimap.
     * @param {*} source - A list of key/value pairs to inject in the map.
     */
    constructor(source) {
        super(source);
    }

    /**
     * Tells whether a key exists or not. When the value is supplied, it also tells whether the key/pair exists or not.
     * @param {*} key - The key to check.
     * @param {*} [value] - An optional value to check together with the key.
     * @returns {boolean} - Returns `true` if the keys exists.
     */
    has(key, value) {
        if ('undefined' === typeof value) {
            return super.has(key);
        }

        const bucket = this.get(key);
        if (bucket) {
            return bucket.has(value);
        }
        return false;
    }

    /**
     * Sets a value under a key. Several different values can be set under the same key. However, each value must be
     * unique within a key.
     * When setting a value to a key, a bucket is created if needed, and the value is added to it. Each further set
     * to the same key will add values to the bucket.
     * @param {*} key - The key to set.
     * @param {*} value - The value to register under the key.
     * @returns {Multimap} - Chains the instance.
     */
    // @ts-expect-error
    set(key, value) {
        let bucket = this.get(key);
        if (!bucket) {
            bucket = new Set();
            super.set(key, bucket);
        }
        bucket.add(value);
        return this;
    }

    /**
     * Deletes a key.
     * If only the key is supplied, the bucket is deleted, removing all registered values as well.
     * If a value is supplied, it is removed from the bucket bound to the key.
     * If the bucket is empty, the key is also removed.
     * @param {*} key - The key to delete.
     * @param {*} [value] - An optional value to remove from the key.
     * @returns {boolean} - Returns `true` if something was deleted.
     */
    delete(key, value) {
        if ('undefined' !== typeof value) {
            const bucket = this.get(key);
            if (bucket) {
                const deleted = bucket.delete(value);
                if (deleted) {
                    super.set(key, bucket);
                }

                if (bucket.size) {
                    return deleted;
                }
            }
        }

        return super.delete(key);
    }

    /**
     * Calls a function on every key/value pairs of the map.
     * @param {eachKeyValue} each - A callback called for each key/value pair.
     * @param {*} [thisArg] - An optional call context for the callback. If omitted it will be defaulted to `undefined`.
     * @throws {TypeError} - If the callback is not supplied or is not a function.
     */
    // @ts-expect-error
    forEach(each, thisArg) {
        validateCallback(each);

        for (const [key, bucket] of super.entries()) {
            for (const value of bucket) {
                each.call(thisArg, value, key, this);
            }
        }
    }

    /**
     * Produces an iterator to loop over all the values. This iterator will give an access to each single value. If a
     * key has several values, each of them will be iterated separately.
     * Say if we have a multimap such as `{ one: [1, 2], two:[3, 4] }`, the iterator will successively loop over the
     * values: `[1, 2, 3, 4]`.
     * @yields {*} - The next value in the map.
     * @generator
     */
    *values() {
        for (const bucket of super.values()) {
            for (const value of bucket) {
                yield value;
            }
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs. This iterator will give an access to each single
     * key/value pair. If a key have several values, each of them will be iterated separately together with its key.
     * Say if we have a multimap such as `{ one: [1, 2], two:[3, 4] }`, the iterator will successively loop over the
     * entries: `[['one', 1], ['one', 2], ['two', 3], ['two', 4]]`.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    // @ts-expect-error
    *entries() {
        for (const [key, bucket] of super.entries()) {
            for (const value of bucket) {
                yield [key, value];
            }
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs. This iterator will give an access to each single
     * key/value pair. If a key have several values, each of them will be iterated separately together with its key.
     * Say if we have a multimap such as `{ one: [1, 2], two:[3, 4] }`, the iterator will successively loop over the
     * entries: `[['one', 1], ['one', 2], ['two', 3], ['two', 4]]`.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    // @ts-expect-error
    *[Symbol.iterator]() {
        yield* this.entries();
    }
}

/**
 * Represents a callback function that is invoked for each key/value pair of a `Multimap` object.
 * Mainly used when applying a traversal algorithm, like `forEach()`.
 * @param {*} value - The current value being traversed.
 * @param {*} key - The current key being traversed.
 * @param {Multimap} map - The `Multimap` instance being traversed.
 * @callback eachKeyValue
 */
