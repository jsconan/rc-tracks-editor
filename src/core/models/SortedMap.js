/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateCallback } from '../helpers';
import { TreeNode } from './TreeNode.js';

/**
 * Defines a sorted map structure.
 *
 * A sorted map is a map where the keys are sorted so that a traversal will return a sorted list.
 */
export class SortedMap {
    /**
     * Creates a sorted map.
     * @param {*} source - A list of key/value pairs to inject in the map.
     */
    constructor(source) {
        this.tree = TreeNode.create();
        this.size = 0;

        if (source) {
            for (const [key, value] of source) {
                this.set(key, value);
            }
        }
    }

    /**
     * Tells whether a key exists or not.
     * @param {*} key - The key to check.
     * @returns {boolean} - Returns `true` if the keys exist.
     */
    has(key) {
        return this.tree.has(key);
    }

    /**
     * Gets the value registered under the given key.
     * If the value does not exist, `undefined` will be returned.
     * @param {*} key - The key to check.
     * @returns {*} - Returns the registered value, or `undefined`.
     */
    get(key) {
        return this.tree.get(key);
    }

    /**
     * Sets a value under a key.
     * @param {*} key - The key to set.
     * @param {*} value - The value to register under the key.
     * @returns {SortedMap} - Chains the instance.
     */
    set(key, value) {
        const feedback = {};
        this.tree = this.tree.insert(key, value, feedback);

        if (feedback.added) {
            this.size++;
        }

        return this;
    }

    /**
     * Deletes a key.
     * @param {*} key - The key to remove.
     * @returns {boolean} - Returns `true` if something was deleted.
     */
    delete(key) {
        const feedback = {};
        this.tree = this.tree.delete(key, feedback);

        if (feedback.deleted) {
            this.size--;
            return true;
        }

        return false;
    }

    /**
     * Removes all elements from the map.
     */
    clear() {
        this.tree = TreeNode.create();
        this.size = 0;
    }

    /**
     * Calls a function on every key/value pairs of the map.
     * @param {eachKeyValue} each - A callback called for each keys/value pair.
     * @param {*} [thisArg] - An optional call context for the callback. If omitted it will be defaulted to `undefined`.
     * @throws {TypeError} - If the callback is not supplied or is not a function.
     */
    forEach(each, thisArg) {
        validateCallback(each);

        for (const { key, value } of this.tree) {
            each.call(thisArg, value, key, this);
        }
    }

    /**
     * Produces an iterator to loop over all the keys.
     * @yields {*} - The next key in the map.
     * @generator
     */
    *keys() {
        for (const { key } of this.tree) {
            yield key;
        }
    }

    /**
     * Produces an iterator to loop over all the values.
     * @yields {*} - The next value in the map.
     * @generator
     */
    *values() {
        for (const { value } of this.tree) {
            yield value;
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    *entries() {
        for (const { key, value } of this.tree) {
            yield [key, value];
        }
    }

    /**
     * Produces an iterator to loop over all the key/value pairs.
     * @yields {*} - The next key/value pair in the map.
     * @generator
     */
    *[Symbol.iterator]() {
        yield* this.entries();
    }
}

/**
 * Represents a callback function that is invoked for each key/value pair of a `SortedMap` object.
 * Mainly used when applying a traversal algorithm, like `forEach()`.
 * @param {*} value - The current value being traversed.
 * @param {*} key - The current key being traversed.
 * @param {SortedMap} map - The `SortedMap` instance being traversed.
 * @returns {*} - It may return a value that will impact the result of the visit.
 * @callback eachKeyValue
 */
