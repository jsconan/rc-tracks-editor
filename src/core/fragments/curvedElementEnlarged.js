/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { SVGPath, Vector2D } from '../models';

/**
 * Builds the SVG path for rendering an enlarged curve.
 * @param {number} x - The X-coordinate of the center of the curve.
 * @param {number} y - The Y-coordinate of the center of the curve.
 * @param {number} width - The width between the inner and the outer curve.
 * @param {number} radius - The radius of the inner curve.
 * @param {number} side - Tha length of the curve side.
 * @param {number} addition - An additional distance added to the outline.
 * @returns {SVGPath} - Returns the path for the shape.
 */
export function curvedElementEnlargedPath(x, y, width, radius, side, addition = 0) {
    addition = Math.min(addition, radius);

    const outerRadius = radius + width - side + addition;
    const center = new Vector2D(x, y).subScalar(addition);

    return new SVGPath()
        .moveTo(center.addScalarX(radius))
        .ellipticalArcCurveTo(radius, 0, 0, 1, center.addScalarY(radius))
        .verticalLineBy(width + addition * 2)
        .horizontalLineBy(side + addition)
        .ellipticalArcCurveBy(outerRadius, 0, 0, 0, new Vector2D(outerRadius, -outerRadius))
        .verticalLineBy(-side - addition)
        .close();
}
