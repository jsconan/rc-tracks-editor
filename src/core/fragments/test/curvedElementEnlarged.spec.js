/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { curvedElementEnlargedPath } from '../curvedElementEnlarged.js';
import { SVGPath } from '../../models';

describe('curvedElementEnlargedPath', () => {
    it('is a function', () => {
        expect(curvedElementEnlargedPath).toEqual(expect.any(Function));
    });

    it('returns a SVG path', () => {
        expect(curvedElementEnlargedPath(0, 0, 10, 20, 5)).toBeInstanceOf(SVGPath);
    });

    it.each([
        [100, 100, 60, 50, 20, void 0],
        [100, 100, 60, 50, 20, 5]
    ])(
        'draws a curved arrow tip at %s,%s having width=%s, radius=%s and side=%s, with %s added',
        (x, y, width, radius, side, addition) => {
            expect(curvedElementEnlargedPath(x, y, width, radius, side, addition).toString()).toMatchSnapshot();
        }
    );
});
