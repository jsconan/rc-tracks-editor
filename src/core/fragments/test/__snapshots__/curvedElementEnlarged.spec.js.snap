// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`curvedElementEnlargedPath draws a curved arrow tip at 100,100 having width=60, radius=50 and side=20, with 5 added 1`] = `"M 145,95 A 50,50 0 0 1 95,145 V 215 H 120 A 95,95 0 0 0 215,120 V 95 Z"`;

exports[`curvedElementEnlargedPath draws a curved arrow tip at 100,100 having width=60, radius=50 and side=20, with undefined added 1`] = `"M 150,100 A 50,50 0 0 1 100,150 V 210 H 120 A 90,90 0 0 0 210,120 V 100 Z"`;
